### PR TITLE
Implement duckdb scanner extension

### PR DESCRIPTION
**Pull request by acquamarin**
This is the first set of PR that implements duckdb scanner. 
Duckdb scanner allows user to scan duckdb tables in kuzu.

### CODE REVIEW

File: src/include/processor/operator/attach_database.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    bool getNextTuplesInternal(ExecutionContext* context) override;
+
+    std::unique_ptr<PhysicalOperator> clone() override {
+        return std::make_unique<AttachDatabase>(attachInfo, id, paramsString);
+    }
+
+private:
+    parser::AttachInfo attachInfo;
+    bool hasExecuted = false;
```
Comment (andyfengHKU) [inline src/include/processor/operator/attach_database.h:26]: remove

File: src/include/processor/operator/detach_database.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    bool getNextTuplesInternal(ExecutionContext* context) override;
+
+    std::unique_ptr<PhysicalOperator> clone() override {
+        return std::make_unique<DetachDatabase>(dbName, id, paramsString);
+    }
+
+private:
+    std::string dbName;
+    bool hasExecuted = false;
```
Comment (andyfengHKU) [inline src/include/processor/operator/detach_database.h:25]: remove

File: src/include/processor/operator/attach_database.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+namespace processor {
+
+class AttachDatabase : public PhysicalOperator {
+public:
+    AttachDatabase(parser::AttachInfo attachInfo, uint32_t id, const std::string& paramsString)
+        : PhysicalOperator{PhysicalOperatorType::ATTACH_DATABASE, id, paramsString},
+          attachInfo{std::move(attachInfo)} {}
+
+    bool isSource() const override { return true; }
+    bool canParallel() const final { return false; }
```
Comment (andyfengHKU) [inline src/include/processor/operator/attach_database.h:16]: override and mark class as final

File: src/include/processor/operator/detach_database.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+namespace processor {
+
+class DetachDatabase : public PhysicalOperator {
+public:
+    DetachDatabase(std::string dbName, uint32_t id, const std::string& paramsString)
+        : PhysicalOperator{PhysicalOperatorType::DETACH_DATABASE, id, paramsString},
+          dbName{std::move(dbName)} {}
+
+    bool isSource() const override { return true; }
+    bool canParallel() const final { return false; }
```
Comment (andyfengHKU) [inline src/include/processor/operator/detach_database.h:15]: override and mark class as final

File: src/processor/operator/physical_operator.cpp [hunk 5]
```diff
@@ -15,6 +15,8 @@ std::string PhysicalOperatorUtils::operatorTypeToString(PhysicalOperatorType ope
         return "AGGREGATE";
     case PhysicalOperatorType::AGGREGATE_SCAN:
         return "AGGREGATE_SCAN";
+    case PhysicalOperatorType::ATTACH_DATABASE:
```
Comment (andyfengHKU) [inline src/processor/operator/physical_operator.cpp:18]: DETACH is missing

File: src/processor/operator/detach_database.cpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace kuzu {
+namespace processor {
+
+bool DetachDatabase::getNextTuplesInternal(kuzu::processor::ExecutionContext* context) {
+    if (hasExecuted) {
+        return false;
+    }
+    hasExecuted = true;
+    auto dbManager = context->clientContext->getDatabase()->getDatabaseManagerUnsafe();
```
Comment (andyfengHKU) [inline src/processor/operator/detach_database.cpp:14]: remove l10 - l13

File: src/processor/operator/attach_database.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#include "storage/storage_extension.h"
+
+namespace kuzu {
+namespace processor {
+
+bool AttachDatabase::getNextTuplesInternal(kuzu::processor::ExecutionContext* context) {
+    if (hasExecuted) {
+        return false;
+    }
+    hasExecuted = true;
```
Comment (andyfengHKU) [inline src/processor/operator/attach_database.cpp:14]: remove l11-l14

File: src/include/storage/storage_extension.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include "main/attached_database.h"
+
+namespace kuzu {
+namespace storage {
+
+typedef std::unique_ptr<main::AttachedDatabase> (*attach_function_t)(
```
Comment (andyfengHKU) [inline src/include/storage/storage_extension.h:8]: let use `using`

File: src/include/planner/planner.h [hunk 9]
```diff
@@ -63,10 +65,13 @@ class Planner {
     std::unique_ptr<LogicalPlan> planCopyRdfFrom(
         const binder::BoundCopyFromInfo* info, binder::expression_vector outExprs);
 
-    // Plan export database
+    // Plan export/import database
     std::unique_ptr<LogicalPlan> planExportDatabase(const binder::BoundStatement& statement);
     std::unique_ptr<LogicalPlan> planImportDatabase(const binder::BoundStatement& statement);
 
+    // Plan attach/detach database
+    std::unique_ptr<LogicalPlan> planAttachDatabase(const binder::BoundStatement& statement);
```
Comment (andyfengHKU) [inline src/include/planner/planner.h:73]: is this needed?

File: src/include/main/database_manager.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "attached_database.h"
+
+namespace kuzu {
+namespace main {
+
+class DatabaseManager {
+public:
+    void registerAttachedDatabase(std::unique_ptr<AttachedDatabase> attachedDatabase);
+    AttachedDatabase* getAttachedDatabase(std::string name);
```
Comment (andyfengHKU) [inline src/include/main/database_manager.h:11]: I suggest we do `const std::string&` when possible

File: src/include/main/database.h [hunk 11]
```diff
@@ -131,6 +145,8 @@ class Database {
     std::shared_ptr<spdlog::logger> logger;
     std::unique_ptr<common::FileInfo> lockFile;
     std::unique_ptr<extension::ExtensionOptions> extensionOptions;
+    std::unique_ptr<DatabaseManager> databaseManager;
```
Comment (andyfengHKU) [inline src/include/main/database.h:148]: Let's use object for DatabaseManager and StorageExtension if possible

File: src/include/main/attached_database.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+class AttachedDatabase {
+
+public:
+    AttachedDatabase(std::string dbName, function::TableFunction scanFunction)
+        : dbName{std::move(dbName)}, scanFunction{std::move(scanFunction)} {}
+
+    std::string getDBName() { return dbName; }
+
+    function::TableFunction* getScanFunction() { return &scanFunction; }
```
Comment (andyfengHKU) [inline src/include/main/attached_database.h:18]: I wonder if it is better to just copy TableFunction?

File: src/include/function/table_functions.h [hunk 13]
```diff
@@ -64,8 +64,8 @@ struct TableFunctionInitInput {
     virtual ~TableFunctionInitInput() = default;
 };
 
-typedef std::unique_ptr<TableFuncBindData> (*table_func_bind_t)(
-    main::ClientContext* context, TableFuncBindInput* input);
+using table_func_bind_t = std::function<std::unique_ptr<function::TableFuncBindData>(
```
Comment (andyfengHKU) [inline src/include/function/table_functions.h:67]: remove `function::`

File: src/include/common/vector/value_vector.h [hunk 14]
```diff
@@ -17,7 +17,7 @@ class Value;
 
 //! A Vector represents values of the same data type.
 //! The capacity of a ValueVector is either 1 (sequence) or DEFAULT_VECTOR_CAPACITY.
-class ValueVector {
+class KUZU_API ValueVector {
```
Comment (andyfengHKU) [inline src/include/common/vector/value_vector.h:20]: Double check with @mewim for KUZU_API over class

File: src/binder/bind/bind_reading_clause.cpp [hunk 15]
```diff
@@ -157,15 +160,29 @@ std::unique_ptr<BoundReadingClause> Binder::bindLoadFrom(const ReadingClause& re
...
-        auto objectExpr = expressionBinder.bindVariableExpression(objectName);
-        auto literalExpr =
-            ku_dynamic_cast<const Expression*, const LiteralExpression*>(objectExpr.get());
-        auto functions = clientContext->getCatalog()->getFunctions(clientContext->getTx());
-        auto func = BuiltInFunctionsUtils::matchFunction(
-            READ_PANDAS_FUNC_NAME, std::vector<LogicalType>{objectExpr->getDataType()}, functions);
-        scanFunction = ku_dynamic_cast<Function*, TableFunction*>(func);
-        bindInput = std::make_unique<function::TableFuncBindInput>();
-        bindInput->inputs.push_back(*literalExpr->getValue());
+        if (objectName.find("_") == std::string::npos) {
```
Comment (andyfengHKU) [inline src/binder/bind/bind_reading_clause.cpp:163]: open an issue for me to resolve scanning from external objects

File: src/include/catalog/catalog.h [hunk 16]
```diff
@@ -83,6 +84,7 @@ class Catalog {
     function::ScalarMacroFunction* getScalarMacroFunction(const std::string& name) const {
         return readOnlyVersion->getScalarMacroFunction(name);
     }
+    void createTableFunction(std::string name, function::function_set tableFunction);
```
Comment (andyfengHKU) [inline src/include/catalog/catalog.h:87]: This doesn't seem to be needed.

File: extension/duckdb_scanner/src/include/duckdb_scan.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+struct DuckDBScanSharedState : public function::TableFuncSharedState {
+    explicit DuckDBScanSharedState(std::unique_ptr<duckdb::QueryResult> queryResult);
+
+    std::unique_ptr<duckdb::QueryResult> queryResult;
+};
+
+struct DuckDBScanFunction {
+    static constexpr char DUCKDB_SCAN_FUNC_NAME[] = "duckdb_scan";
+
+    static std::unique_ptr<function::TableFunction> getScanFunction(std::string dbPath);
```
Comment (andyfengHKU) [inline extension/duckdb_scanner/src/include/duckdb_scan.h:41]: I think you only need to expose this function and can move the rest to cpp file

**Review by andyfengHKU — APPROVED** [review]
LGTM

File: src/include/planner/planner.h (hunk 9 above)
Comment (acquamarin) [reply src/include/planner/planner.h:73]: removed

File: src/include/main/database.h (hunk 11 above)
Comment (acquamarin) [reply src/include/main/database.h:148]: If we store the objet for DatabaseManager and StorageExtension we have to include the headers in the database.h.

File: src/include/main/attached_database.h (hunk 12 above)
Comment (acquamarin) [reply src/include/main/attached_database.h:18]: We can avoid an additional copy if we return by pointer and the funciton implementation is always the same.

**Pull request merged**